src/pemfile.rs

```diff
@@ -37,6 +43,56 @@ pub enum Item {
     Crl(CertificateRevocationListDer<'static>),
 }
 
+/// Errors that may arise when parsing the contents of a PEM file
+pub enum Error {
+    /// a section is missing its "END marker" line
+    MissingSectionEnd {
+        /// the expected "END marker" line that was not found
+        end_marker: Vec<u8>,
+    },
+
+    /// syntax error found in the line that starts a new section
+    IllegalSectionStart {
+        /// line that contains the syntax error
+        line: Vec<u8>,
+    },
+
+    /// base64 decode error
+    Base64Decode(DecodeError),
+}
+
+impl From<DecodeError> for Error {
```

Comment:
This leaks the `base64` crate into the public API no? I think one advantage of the previous approach was that it kept `base64` internal.

Reply:
Indeed. we can't use the usual `Box<dyn std::error::Error>` replacement because this enum has to work in no-std context. the alternatives I see are

- drop the tuple field, i.e. simply use `Error::Base64Decode`. this loses information
- use `Base64Decode(Box<dyn core::fmt::Debug>)`. that trait object is not super common but it's close to `Box<dyn Error>`. however, it cannot be converted into `Box<dyn Error>` as there's no `From` / `Into` trait that glues the two trait object types
- use `Base64Decode(String)` where the `String` is the `format!`-ed version of `Base64Decode`. unlike `Box<dyn Debug>` this can be converted into `Box<dyn Error>` using the `?` operator so that's a plus
- basically copy `base64::DecodeError` into this crate and use that. that does not expose the `base64` dependency but then you have to maintain a function that maps `base64::DecodeError` into `pemfile::DecodeError` and update it if `base64::DecodeError` (which is not marked as non_exhaustive) ever changes.

Reply:
@ctz and I yesterday were chatting that we might want to yeet the use of the base64 crate at the first sign of trouble (since we generally haven't been that happy with it) -- maybe that's this?

Reply:
> use Base64Decode(String) where the String is the format!-ed version of Base64Decode. unlike Box<dyn Debug> this can be converted into Box<dyn Error> using the ? operator so that's a plus

That option seems OK to me. I suspect there's not much value in downstream code being able to programmatically determine if the data was invalid for padding, length, unknown symbol etc. In most cases I expect the error is going to be displayed to a user in string form :thinking:

Reply:
>  In most cases I expect the error is going to be displayed to a user in string form

I agree. this is not the case where knowing the exact error cause is going to let the user fix the input and retry the operation. most likely, malformed pem files are going to be skipped or result in a fatal error.